### PR TITLE
embed: only override default advertised client URL if the client listen URL is 0.0.0.0

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/7009.

### CODE REVIEW

File: embed/config.go [hunk 1]
```diff
@@ -346,14 +331,27 @@ func (cfg Config) ElectionTicks() int { return int(cfg.ElectionMs / cfg.TickMs)
...
-		return defaultHostname, defaultHostStatus
+func (cfg *Config) IsDefaultHost() (string, error) {
+	h, herr := netutil.GetDefaultHost()
+	if len(cfg.APUrls) == 1 {
+		ip, port, err := net.SplitHostPort(cfg.APUrls[0].Host)
+		if err != nil {
+			return "", err
+		}
+		if ip == "0.0.0.0" {
+			cfg.APUrls[0] = url.URL{Scheme: cfg.APUrls[0].Scheme, Host: fmt.Sprintf("%s:%s", h, port)}
```
Comment (heyitsanthony) [inline embed/config.go]: This function is misleading now; usually `Is` functions are idempotent. Here, `cfg.IsDefaultHost()` won't necessarily return the same value on subsequent calls. Probably should just have a separate function do to the advertise IP override.

File: embed/config.go [hunk 2]
```diff
@@ -346,14 +331,27 @@ func (cfg Config) ElectionTicks() int { return int(cfg.ElectionMs / cfg.TickMs)
 
 // IsDefaultHost returns the default hostname, if used, and the error, if any,
 // from getting the machine's default host.
-func (cfg Config) IsDefaultHost() (string, error) {
-	if len(cfg.APUrls) == 1 && cfg.APUrls[0].String() == DefaultInitialAdvertisePeerURLs {
-		return defaultHostname, defaultHostStatus
+func (cfg *Config) IsDefaultHost() (string, error) {
+	h, herr := netutil.GetDefaultHost()
+	if len(cfg.APUrls) == 1 {
```
Comment (heyitsanthony) [inline embed/config.go]: The 0.0.0.0 check should be on the listen url. If 0.0.0.0, then the advertise urls need to be checked if they match the default url before overriding.

**Comment by gyuho** [comment]
@heyitsanthony PTAL.

```
./bin/etcd
2016-12-16 13:46:36.571159 I | embed: listening for peers on http://localhost:2380
2016-12-16 13:46:36.571222 I | embed: listening for client requests on localhost:2379
...
2016-12-16 13:46:36.573351 I | etcdserver: advertise client URLs = http://localhost:2379


./bin/etcd --listen-peer-urls=http://0.0.0.0:2380

...
2016-12-16 13:43:06.344098 I | etcdmain: advertising using detected default host "10.7.3.90"
2016-12-16 13:43:06.344141 I | embed: listening for peers on http://0.0.0.0:2380
2016-12-16 13:43:06.344306 I | embed: listening for client requests on localhost:2379
...
2016-12-16 13:43:06.350580 I | etcdserver: advertise client URLs = http://10.7.3.90:2379
2016-12-16 13:43:06.350587 I | etcdserver: initial advertise peer URLs = http://10.7.3.90:2380
2016-12-16 13:43:06.350599 I | etcdserver: initial cluster = default=http://10.7.3.90:2380
```

Thanks!

File: embed/config.go [hunk 3]
```diff
@@ -344,16 +328,25 @@ func (cfg Config) InitialClusterFromName(name string) (ret string) {
...
+	if len(cfg.LCUrls) == 1 {
+		ip, _, _ := net.SplitHostPort(cfg.LCUrls[0].Host)
+		if ip == "0.0.0.0" {
+			if len(cfg.ACUrls) == 1 && cfg.ACUrls[0].String() == DefaultAdvertiseClientURLs {
+				return true
+			}
+		}
 	}
-	return "", defaultHostStatus
+	return false
```
Comment (heyitsanthony) [inline embed/config.go]: this changes the meaning of the function though? This function was meant to say whether it's using the default hosts /  user didn't specify, but now it's returning true if it's overriding the hosts.

File: etcdmain/etcd.go [hunk 4]
```diff
@@ -85,10 +87,21 @@ func startEtcdOrProxyV2() {
 	plog.Infof("setting maximum number of CPUs to %d, total number of available CPUs is %d", GoMaxProcs, runtime.NumCPU())
 
 	// TODO: check whether fields are set instead of whether fields have default value
-	defaultHost, defaultHostErr := cfg.IsDefaultHost()
-	defaultHostOverride := defaultHost == "" || defaultHostErr == nil
-	if (defaultHostOverride || cfg.Name != embed.DefaultName) && cfg.InitialCluster == defaultInitialCluster {
-		cfg.InitialCluster = cfg.InitialClusterFromName(cfg.Name)
+	needDefaultHost := cfg.IsDefaultHost()
+	if (needDefaultHost || cfg.Name != embed.DefaultName) && cfg.InitialCluster == defaultInitialCluster {
+		defaultHost, dhErr := netutil.GetDefaultHost()
```
Comment (heyitsanthony) [inline etcdmain/etcd.go]: this overriding stuff should be a function in embed.Config so it can be reused elsewhere

File: embed/config.go (hunk 3 above)
Comment (gyuho) [inline embed/config.go]: You are right. How about `NeedDefaultHost()`, which means it returns `true` if it needs to detect machine's default host?

Comment (heyitsanthony) [inline embed/config.go]: I don't think that will work-- `IsDefaultHost` means that it's using defaults (which may be overridden with the default route IP), not that it's using the default route IP. That will change the meaning of the InitialClusterFromName path, which needs to know whether the user set a host explicitly.

**Comment by gyuho** [comment]
@heyitsanthony Now it does

> only override the default advertised client URL if the client listen URL is 0.0.0.0

in a way that if the default host is overwritten and the listen client URL is not default route (0.0.0.0), it overwrites advertise-client-url with `localhost`, since `embed` package uses, in `init` func, the default host.

Thanks.

File: etcdmain/etcd.go [hunk 5]
```diff
@@ -89,6 +89,11 @@ func startEtcdOrProxyV2() {
 	defaultHostOverride := defaultHost == "" || defaultHostErr == nil
 	if (defaultHostOverride || cfg.Name != embed.DefaultName) && cfg.InitialCluster == defaultInitialCluster {
```
Comment (heyitsanthony) [inline etcdmain/etcd.go]: this path needs a lot of special support from Config and seems self-contained, can everything starting from `IsDefaultHost()` to the end of this `if` be `func (cfg *Config) UpdateDefaultClusterFromName()` instead?

Comment (gyuho) [inline etcdmain/etcd.go]: @heyitsanthony Just moved them all to `embed.UpdateDefaultClusterFromName`. PTAL. Thanks.

File: embed/config.go [hunk 6]
```diff
@@ -356,6 +356,32 @@ func (cfg Config) IsDefaultHost() (string, error) {
...
+		cfg.InitialCluster = cfg.InitialClusterFromName(cfg.Name)
+		// if client-listen-url is 0.0.0.0, just use detected default host
+		// otherwise, rewrite advertise-client-url with localhost
+		if !isDefaultRoute(cfg.LCUrls[0].Host) {
+			cfg.updateACURLHost("localhost")
+		}
+	}
+}
+
+func isDefaultRoute(host string) bool {
```
Comment (heyitsanthony) [inline embed/config.go]: can probably inline this instead of a separate function?

File: embed/config.go [hunk 7]
```diff
@@ -356,6 +356,32 @@ func (cfg Config) IsDefaultHost() (string, error) {
...
+	ip, _, _ := net.SplitHostPort(host)
+	return ip == "0.0.0.0"
+}
+
+func (cfg *Config) updateACURLHost(host string) {
```
Comment (heyitsanthony) [inline embed/config.go]: same here

**Comment by heyitsanthony** [comment]
lgtm after fixing minor nits

**Comment by gyuho** [comment]
All fixed. Merging after greens. Thanks!

**Pull request merged**